In [1]:
import pandas as pd
import numpy as np


In [2]:
df= pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Datasets /ai_job_dataset1.csv')

In [3]:
df

,job_id,job_title,salary_usd,salary_currency,salary_local,experience_level,employment_type,company_location,company_size,employee_residence,remote_ratio,required_skills,education_required,years_experience,industry,posting_date,application_deadline,job_description_length,benefits_score,company_name
0,AI00001,Data Scientist,219728,USD,219728,EX,PT,Sweden,M,Sweden,0,"Python, Computer Vision, R, Docker",Associate,13,Transportation,2024-09-23,2024-10-31,1132,6.6,TechCorp Inc
1,AI00002,Head of AI,230237,JPY,25326070,EX,PT,Japan,L,Japan,50,"Kubernetes, MLOps, Tableau, Python",Bachelor,10,Transportation,2024-07-26,2024-09-12,2299,8.5,Cloud AI Solutions
2,AI00003,Data Engineer,128890,EUR,109557,EX,CT,Germany,S,Germany,100,"Spark, Scala, Hadoop, PyTorch, GCP",Bachelor,12,Automotive,2025-01-19,2025-03-28,1329,5.5,Quantum Computing Inc
3,AI00004,Computer Vision Engineer,96349,USD,96349,MI,FL,Finland,L,Finland,50,"MLOps, Linux, Tableau, Python",PhD,2,Automotive,2024-07-20,2024-09-06,1132,6.8,Cognitive Computing
4,AI00005,Robotics Engineer,63065,EUR,53605,EN,FT,France,S,France,100,"R, Scala, SQL, GCP, Python",Associate,0,Finance,2025-03-16,2025-05-09,2011,9.3,Advanced Robotics
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14995,AI14996,AI Product Manager,39171,USD,39171,MI,CT,China,M,China,50,"Azure, R, NLP, Docker, Computer Vision",PhD,4,Consulting,2025-01-07,2025-02-07,1107,5.2,Smart Analytics
14996,AI14997,AI Product Manager,77555,EUR,65922,EN,CT,Netherlands,M,Netherlands,50,"Python, TensorFlow, Mathematics, AWS, Computer...",PhD,0,Energy,2024-12-22,2025-02-12,1870,9.7,Autonomous Tech
14997,AI14998,Data Engineer,28380,USD,28380,EN,FL,India,M,India,50,"Azure, Kubernetes, Spark, Statistics, MLOps",Associate,1,Government,2024-05-04,2024-05-21,1558,9.0,Digital Transformation LLC
14998,AI14999,AI Specialist,58764,EUR,49949,EN,PT,France,M,France,100,"MLOps, Statistics, Data Visualization, R, Python",Master,0,Real Estate,2024-05-08,2024-06-22,546,8.4,Machine Intelligence Group


# EDA Preprocessing

In [4]:
df.head()

,job_id,job_title,salary_usd,salary_currency,salary_local,experience_level,employment_type,company_location,company_size,employee_residence,remote_ratio,required_skills,education_required,years_experience,industry,posting_date,application_deadline,job_description_length,benefits_score,company_name
0,AI00001,Data Scientist,219728,USD,219728,EX,PT,Sweden,M,Sweden,0,"Python, Computer Vision, R, Docker",Associate,13,Transportation,2024-09-23,2024-10-31,1132,6.6,TechCorp Inc
1,AI00002,Head of AI,230237,JPY,25326070,EX,PT,Japan,L,Japan,50,"Kubernetes, MLOps, Tableau, Python",Bachelor,10,Transportation,2024-07-26,2024-09-12,2299,8.5,Cloud AI Solutions
2,AI00003,Data Engineer,128890,EUR,109557,EX,CT,Germany,S,Germany,100,"Spark, Scala, Hadoop, PyTorch, GCP",Bachelor,12,Automotive,2025-01-19,2025-03-28,1329,5.5,Quantum Computing Inc
3,AI00004,Computer Vision Engineer,96349,USD,96349,MI,FL,Finland,L,Finland,50,"MLOps, Linux, Tableau, Python",PhD,2,Automotive,2024-07-20,2024-09-06,1132,6.8,Cognitive Computing
4,AI00005,Robotics Engineer,63065,EUR,53605,EN,FT,France,S,France,100,"R, Scala, SQL, GCP, Python",Associate,0,Finance,2025-03-16,2025-05-09,2011,9.3,Advanced Robotics


In [5]:
df.shape

(15000, 20)

There are 15000 rows and 20 colns

In [6]:
df.dtypes

,0
job_id,object
job_title,object
salary_usd,int64
salary_currency,object
salary_local,int64
experience_level,object
employment_type,object
company_location,object
company_size,object
employee_residence,object


there are total 14 categorical ,6 interger values features

In [7]:
df.isnull().sum()

,0
job_id,0
job_title,0
salary_usd,0
salary_currency,0
salary_local,0
experience_level,0
employment_type,0
company_location,0
company_size,0
employee_residence,0


Thare are NO missing values


In [8]:
df.columns

Index(['job_id', 'job_title', 'salary_usd', 'salary_currency', 'salary_local',
       'experience_level', 'employment_type', 'company_location',
       'company_size', 'employee_residence', 'remote_ratio', 'required_skills',
       'education_required', 'years_experience', 'industry', 'posting_date',
       'application_deadline', 'job_description_length', 'benefits_score',
       'company_name'],
      dtype='object')

In [9]:
columns_to_keep =['job_title',
'salary_usd',
'experience_level',
'employment_type',
'company_location',
'remote_ratio',
'required_skills',
'education_required',
'years_experience',
'industry',
'company_size',
'posting_date'
]
df=df[columns_to_keep]

In [10]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [11]:
for i in df:
  if (df[i].dtype=='object'):
    df[i]=le.fit_transform(df[i])

/tmp/ipython-input-845966945.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i]=le.fit_transform(df[i])
/tmp/ipython-input-845966945.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i]=le.fit_transform(df[i])
/tmp/ipython-input-845966945.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.htm

In [12]:
df.drop(columns=['posting_date'], inplace=True)


/tmp/ipython-input-3009665090.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=['posting_date'], inplace=True)


In [13]:
df.head()

,job_title,salary_usd,experience_level,employment_type,company_location,remote_ratio,required_skills,education_required,years_experience,industry,company_size
0,10,219728,1,3,16,0,8316,0,13,14,1
1,12,230237,1,3,11,50,5232,1,10,14,0
2,9,128890,1,0,7,100,12258,1,12,0,2
3,7,96349,2,1,5,50,6517,3,2,0,0
4,19,63065,0,2,6,100,10033,0,0,4,2


In [14]:
Target= 'salary_usd'

In [15]:
num_coln = [
    'years_experience',
    'remote_ratio'
]

In [16]:
X = df.drop('salary_usd', axis=1)
y = df['salary_usd']

In [17]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2 , random_state=1)

In [18]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train[num_coln] = scaler.fit_transform(X_train[num_coln])
X_test[num_coln] = scaler.transform(X_test[num_coln])

In [19]:
df.describe()

,job_title,salary_usd,experience_level,employment_type,company_location,remote_ratio,required_skills,education_required,years_experience,industry,company_size
count,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000,15000.000000
mean,9.515000,121991.938267,1.506267,1.491333,9.600067,50.196667,6862.478667,1.501467,6.365667,6.956933,0.992533
std,5.756178,63968.361846,1.111655,1.123866,5.783409,40.844084,3916.684362,1.115025,5.598551,4.309747,0.819017
min,0.000000,16621.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,5.000000,74978.500000,1.000000,0.000000,5.000000,0.000000,3486.750000,1.000000,2.000000,3.000000,0.000000
50%,9.000000,107261.500000,2.000000,1.000000,10.000000,50.000000,6908.500000,1.000000,5.000000,7.000000,1.000000
75%,14.000000,155752.250000,2.000000,3.000000,15.000000,100.000000,10216.000000,3.000000,10.000000,11.000000,2.000000
max,19.000000,410273.000000,3.000000,3.000000,19.000000,100.000000,13650.000000,3.000000,19.000000,14.000000,2.000000


1. Average Salary_usd is 121991
2. And min value of slary_usd is 16621 and max is  410273

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   job_title           15000 non-null  int64
 1   salary_usd          15000 non-null  int64
 2   experience_level    15000 non-null  int64
 3   employment_type     15000 non-null  int64
 4   company_location    15000 non-null  int64
 5   remote_ratio        15000 non-null  int64
 6   required_skills     15000 non-null  int64
 7   education_required  15000 non-null  int64
 8   years_experience    15000 non-null  int64
 9   industry            15000 non-null  int64
 10  company_size        15000 non-null  int64
dtypes: int64(11)
memory usage: 1.3 MB


# Model Selection


## Linear Regression

In [21]:
from sklearn.linear_model import LinearRegression

li = LinearRegression()

In [22]:
li.fit(X_train,y_train)

LinearRegression()

In [23]:
li.intercept_

np.float64(116523.58976905124)

The value of c is 116523 --->(y= mX+c)

In [24]:
li.coef_

array([ 4.27670187e+01,  7.79692448e+02,  5.84183284e+02,  1.85210892e+03,
       -1.49038988e+02,  4.82177105e-02, -4.50156443e+02,  4.76713418e+04,
        4.82499625e+01, -1.47210226e+04])

There are 10 multiple m values

In [25]:
y_pred = li.predict(X_test)

In [26]:
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score

In [27]:
print("MAE:", mean_absolute_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("R2:", r2_score(y_test, y_pred))

MAE: 27735.23864363301
RMSE: 39497.341708779684
R2: 0.6134791079457446


## Decision Tree

In [28]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import classification_report,accuracy_score

dt = DecisionTreeRegressor()

In [29]:
dt.fit(X_train,y_train)

DecisionTreeRegressor()

In [30]:
y_pred = dt.predict(X_test)

In [31]:
print("MAE:", mean_absolute_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("R2:", r2_score(y_test, y_pred))

MAE: 20917.297666666665
RMSE: 30568.766691515924
R2: 0.7684775693580171


## RandomForest


In [32]:
from sklearn.ensemble import RandomForestRegressor

In [33]:
rf = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

rf.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [34]:
y_pred = rf.predict(X_test)

In [35]:
print("MAE:", mean_absolute_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))
print("R2:", r2_score(y_test, y_pred))

MAE: 16077.360789999999
RMSE: 22665.12300641633
R2: 0.8727219832854679


# FOR POWERBI

In [ ]:
df['experience_level'] = df['experience_level'].replace({
    'EN' :'Entry',
    'MI' :'Mid',
    'SE' :'Senior',
    'EX' :'Executive'
})

/tmp/ipython-input-488732503.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['experience_level'] = df['experience_level'].replace({


In [ ]:
df['employment_type'] = df['employment_type'].replace({
    'FT': 'Full-time',
    'PT': 'Part-time',
    'CT': 'Contract',
    'FL': 'Freelance'
})

/tmp/ipython-input-4260995477.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['employment_type'] = df['employment_type'].replace({


In [ ]:
df['remote_ratio']=df['remote_ratio'].replace({
    0 :'No remote',
    50: 'Hybrid',
    100 :'Fully remote'
})

df['company_size'] = df['company_size'].replace({
    'M' : 'Mid(50-250)',
    'S' : 'Small(<50)',
    'L' : 'Large(>250)'
})

/tmp/ipython-input-3677414527.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['remote_ratio']=df['remote_ratio'].replace({
/tmp/ipython-input-3677414527.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['company_size'] = df['company_size'].replace({


In [ ]:
df.head()

,job_title,salary_usd,experience_level,employment_type,company_location,remote_ratio,required_skills,education_required,years_experience,industry,company_size
0,10,219728,1,3,16,No remote,8316,0,13,14,1
1,12,230237,1,3,11,Hybrid,5232,1,10,14,0
2,9,128890,1,0,7,Fully remote,12258,1,12,0,2
3,7,96349,2,1,5,Hybrid,6517,3,2,0,0
4,19,63065,0,2,6,Fully remote,10033,0,0,4,2


In [ ]:
df.to_csv("cleaned_ai_job_market.csv", index=False)